hash/hash_comparator.go

```diff
@@ -132,12 +141,42 @@ func CompareScrypt(_ context.Context, password []byte, hash []byte) error {
 	return errors.WithStack(ErrMismatchedHashAndPassword)
 }
 
+func CompareSSHA(_ context.Context, password []byte, hash []byte) error {
+
+	if _, err := ssha.Validate(string(password), string(hash)); err != nil {
```

Comment:
The `ssha` library does not do a constant time comparison of the password with the hash. Is this not an issue?

Reply:
Interesting, I hadn't considered this.  Thanks for pointing it out.  I will look into this and mark this in a draft state for now.

Reply:
I've opened https://github.com/jsimonetti/pwscheme/pull/4 to resolve this.  This PR will remain in draft mode for now as I await review there.  Thanks for the direction!

Reply:
Constant time comparison has been merged into the included library.  Setting this back to ready for review!

Reply:
I believe you still need to update the go.mod file to use these latest changes.

Reply:
Sleeping over it I think we should just call directly to `crypto/sha` and friends instead of using a library for the following reasons:

* I really want to stick to the PHC format specified here: https://github.com/P-H-C/phc-string-format/blob/master/phc-sf-spec.md, which means using `$` as delimitiers
* We will need some flexibility in the handling of the salt for hash functions that do not use a salt as part of their spec (such as bcrypt). Applications may have computed the hash by appending or prepending the salt, or by inserting some separator between hash and salt. See the discussion and my comment on a related PR here: https://github.com/ory/kratos/pull/2725#discussion_r977279806. I think we should use some common solution for all non-password hashes (MD5, SHA-family).
* Lastly, the library still seems quite new, and I'd rather just depend on standard Go libraries for something security critical from a software supply chain standpoint. This is not to say that I hugely respect the effort to create a nice library that automatically handles salting in a nice way, even though these hashes should not be used if you get a choice, because they are just too fast to brute-force, if not broken (like MD5).

So in short, I think we should inline the functionality and make the salting more flexible. WDYT?

Reply:
I definitely understand the sentiments around not including newer/lesser known external security libraries and have thought about that some over the past few days.  I'm good with that approach and the other recommendations you made and will get started on understanding and implementing those changes.

For the various SHA and SSHA implementations, is the desire to be more generic and DRY when handling these or is it a valid approach to still have functions for validating each implementation separately?

Reply:
After reading the linked comments and code more carefully I think I understand now.  The functions to handle SHA, SSHA, SHA256, SSHA256, etc. should handle the presence or absence of a salt dynamically and to your point, while there are norms around how these are persisted, the spec is flexible so the implementation should also be in Kratos.  Thanks!

Reply:
Thinking about this a bit more, a real world example is the string `testing123`.   The SSHA512 sum of this string is:

`xPUl/px+1cG55rUH4rzcwxdOIPSB2TingLpiJJumN2xyDWN4Ix1WQG3ihnvHaWUE8MYNkvMi5rf0C9NYixHsE6Yh59M=`

The salt is included in that SSHA512 sum of course.  The actual salt is:

`piHn0w==`

This could be represented using phc format as:

$SSHA512$hash$salt

for example:

`$SSHA512$xPUl/px+1cG55rUH4rzcwxdOIPSB2TingLpiJJumN2xyDWN4Ix1WQG3ihnvHaWUE8MYNkvMi5rf0C9NYixHsE6Yh59M=$piHn0w==`

I'm not sure if this is the right approach... it feels counter-intuitive to append the salt when it is already part of the hash.  I have some code working with the format.  Since we can't be sure of the formatting I think what is captured above will work but wanted to get some other eyes on it before finishing the rest of the code.

Reply:
Thinking on this more, if the main point of this PR is to be able to import passwords from existing data stores, it becomes problematic to assume that the salt is added in any particular way.  If we are constructing a new hash there is no issue but for preexisting hashes the problem is hairy.  This is discussed some in https://github.com/ory/kratos/pull/2725#discussion_r977279806 but there doesn't seem to be a solution that is clear.

Could we discuss what this would look like in PHC format with real examples?

Reply:
@hperl  and @aeneasr any thoughts on the above?

Reply:
If someone whats to read up on standards for storing password hashes:
- [SO: what is the standard for password hash string encoding?](https://stackoverflow.com/a/73598323/3294973)
- [PHC string format](https://github.com/P-H-C/phc-string-format/blob/master/phc-sf-spec.md)
- [PassLib SHA-1 Crypt Formating](https://passlib.readthedocs.io/en/stable/lib/passlib.hash.sha1_crypt.html#passlib.hash.sha1_crypt)

Reply:
PHC formating: `$<id>[$v=<version>][$<param>=<value>(,<param>=<value>)*][$<salt>[$<hash>]]`

So an example using the formating parameter @hperl proposed over in the [MD5 PR](https://github.com/ory/kratos/pull/2725) would look something like this, right?

`$sha-1$f=SALTPASS$<salt>$<hash>`

Where the format string ("SALTPASS"), salt and hash are base64 encoded.

Reply:
yup, that sounds straight forward @elbotho

Reply:
> * Lastly, the library still seems quite new, and I'd rather just depend on standard Go libraries for something security critical from a software supply chain standpoint.

While I completely understand your supply chain argument, I would like to dispute your claim the library is new. In fact, it is even a lot older then kratos is.